p2p/discovery/message.go

```diff
@@ -139,41 +137,42 @@ func (m *findNode) handle(t *udp, from *net.UDPAddr) {
 		SelfID: t.self.ID,
 	}
 
-	t.sendMsg(NeighborsMsg, response, from)
+	t.sendMsg(neighborsMsgType, response, NewNodeWithAddr(m.SelfID, from))
 }
 
-func (m *findNode) send(t *udp, to *net.UDPAddr) {
-	log.Debug("send find msg to: %s", common.BytesToHex(m.to.Bytes()))
+func (m *findNode) send(t *udp) {
+	log.Debug("send find msg to: %s", hexutil.BytesToHex(m.to.ID.Bytes()))
 
 	p := &pending{
 		from: m.to,
-		code: NeighborsMsg,
+		code: neighborsMsgType,
 
 		callback: func(resp interface{}, addr *net.UDPAddr) (done bool) {
-			log.Debug("received neighbors")
-
 			r := resp.(*neighbors)
 
-			log.Debug("received neighbors msg from: %s", common.BytesToHex(r.SelfID.Bytes()))
+			log.Debug("received neighbors msg from: %s", hexutil.BytesToHex(r.SelfID.Bytes()))
 
 			if r.Nodes == nil || len(r.Nodes) == 0 {
 				return true
 			}
 
+			log.Debug("find %d nodes", len(r.Nodes))
+
 			found := false
 			for _, n := range r.Nodes {
-				if n.SelfID == m.to {
+				log.Debug("received node: %s", hexutil.BytesToHex(n.SelfID.Bytes()))
+
+				if n.SelfID == m.QueryID {
 					found = true
 				}
 
 				node := n.ToNode()
-				t.table.addNode(node)
-				t.db.add(node.sha, node)
+				t.addNode(node)
 			}
 
 			if !found {
```

Comment:
any unfinited loop risk? you never find such node.

Reply:
any find will be more closer to the target, it will stop when can't find a more closer node